notifications_utils/field.py

```diff
@@ -130,19 +131,22 @@ def format_match(self, match):
 
     def replace_match(self, match):
         placeholder = Placeholder.from_match(match)
-        replacement = self.values.get(placeholder.name)
-
-        if placeholder.is_conditional() and replacement is not None:
-            return placeholder.get_conditional_body(replacement)
-
         replaced_value = self.get_replacement(placeholder)
+
+        if placeholder.is_conditional() and replaced_value is not None:
+            return placeholder.get_conditional_body(replaced_value)
         if replaced_value is not None:
-            return self.get_replacement(placeholder)
+            return replaced_value
 
         return self.format_match(match)
 
     def get_replacement(self, placeholder):
-        replacement = self.values.get(placeholder.name)
+        replacement_values_dict = {
+            self.sanitizer(k): v for k, v in self.values.items()
+        }
+        replacement_values = InsensitiveDict(replacement_values_dict)
+        replacement = replacement_values.get(placeholder.name)
```

Comment:
This feels slightly weird and wasteful - we're sanitizing all of the values every time we try to retrieve one of them.

I think we might want to have another attribute similar to `self.values` such as`self.sanitized_values`. We can populate that in the `__init__` method once and then just grab from there as needed. Saves a bit of wasted computation, and also provides another clearer place to add some comments/documentation around this slightly weird workaround.

We definitely want to have some good clear comments describing what's happening here because it's not intuitive or ideal.

Reply:
@samuelhwilliams 
I agree it is not ideal. The challenge of this ticket has been implementing a solution that does not break everything else. I had tried the changes you suggested and about 246 tests failed, the Field class is crucial and changes to it's `__init__` method are consequential.
I have come up with another work around, this time I have given `self.values` the ability to create a `sanitized_content`  dict that can be accessed through it(`self.values`) and this is then used  by the `get_replacement` method. This preserved the current workings of the` Field` class, doesn't break hundreds of tests and solves the bug. I have made comments in the code which I hope are clear.